### PR TITLE
feat(api): block overcapacity upload

### PR DESCRIPTION
**Pull request by jsfez**

### CODE REVIEW

File: apps/database/src/models/Account.js [hunk 1]
```diff
@@ -37,4 +37,8 @@ export class Account extends Model {
       },
     };
   }
+
+  get type() {
+    return this.userId ? "user" : "organization";
+  }
```
Comment (gregberge) [inline apps/database/src/models/Account.js]: `virtualAttributes` not needed for that?

Comment (gregberge) [reply apps/database/src/models/Account.js]: Also we should do:

```js
if (this.userId && this.accountId) throw new Error(`Invariant incoherent account type`)
if (this.userId) return "user"
if (this.accountId) return "organization"
throw new Error(`Invariant incoherent account type`)

File: apps/web/src/middlewares/repoAuth.js [hunk 2]
```diff
@@ -1,8 +1,54 @@
...
+  const date = new Date();
+  date.setDate(1);
+  return new Date(date.toISOString().split("T")[0]);
+}
+
+async function getScreenshotConsumption({ userId, organizationId }) {
+  const query = Screenshot.query()
+    .joinRelated("screenshotBucket.repository")
+    .where("screenshotBucket:repository.private", "true")
+    .where("screenshots.createdAt", ">=", getStartOfMonth());
```
Comment (gregberge) [inline apps/web/src/middlewares/repoAuth.js:21]: Use SQL for that, don't rely on JS.

File: apps/web/src/middlewares/repoAuth.js [hunk 3]
```diff
@@ -1,8 +1,54 @@
...
+  const account = await Account.query()
+    .where("userId", repository.userId)
+    .orWhere("organizationId", repository.organizationId)
+    .first();
+
+  const purchase = await Purchase.query()
+    .withGraphFetched("plan")
+    .where("accountId", account.id)
+    .where((query) =>
+      query.whereNull("endDate").orWhere("endDate", ">=", new Date())
```
Comment (gregberge) [inline apps/web/src/middlewares/repoAuth.js:42]: same, use SQL.

File: apps/database/src/models/Account.js [hunk 4]
```diff
@@ -37,4 +40,69 @@ export class Account extends Model {
...
+      .withGraphFetched("plan")
+      .first();
+
+    return purchase || null;
+  }
+
+  async getScreenshotsLimitPerMonth() {
+    const purchase = await this.getActivePurchase();
+    if (purchase) {
+      return purchase.plan.screenshotsLimitPerMonth === -1
```
Comment (gregberge) [inline apps/database/src/models/Account.js:60]: Factorize this logic on the plan.

File: apps/database/src/models/Account.js [hunk 5]
```diff
@@ -37,4 +40,69 @@ export class Account extends Model {
...
+    const purchase = await Purchase.query()
+      .where("accountId", this.id)
+      .where("startDate", "<", "now()")
+      .where((query) =>
+        query.whereNull("endDate").orWhere("endDate", ">=", "now()")
+      )
+      .withGraphFetched("plan")
+      .first();
+
+    return purchase || null;
```
Comment (gregberge) [inline apps/database/src/models/Account.js:54]: ??

**Review by gregberge — APPROVED** [review]
Nothing critical.

File: apps/database/src/models/Account.js (hunk 5 above)
Comment (jsfez) [reply apps/database/src/models/Account.js:54]: - You: ??
- The server: 🔥

**Pull request merged**